errbot/backends/slack.py

```diff
@@ -90,22 +112,58 @@ def _handle_slack_event(self, event):
                 return
 
             msg = Message(event['text'], type_=message_type)
-            msg.frm = Identifier(node=event['channel'], resource=event['user'])
-            msg.to = Identifier(node=event['channel'], resource=self.jid.node)
+            msg.frm = Identifier(
+                node=self.userid_to_username(event['user']),
+                domain=self.channelid_to_channelname(event['channel'])
+            )
+            msg.to = Identifier(
+                node=self.sc.server.username,
+                domain=self.channelid_to_channelname(event['channel'])
+            )
             self.callback_message(msg)
 
+    def userid_to_username(self, id):
+        """Convert a Slack user ID to their user name"""
+        return self.sc.server.users.find(id).name
+
+    def username_to_userid(self, name):
+        """Convert a Slack user name to their user ID"""
+        return self.sc.server.users.find(name).id
+
+    def channelid_to_channelname(self, id):
+        """Convert a Slack channel ID to its channel name"""
+        return self.sc.server.channels.find(id).name
+
+    def channelname_to_channelid(self, name):
+        """Convert a Slack channel name to its channel ID"""
+        return self.sc.server.channels.find(name).id
+
     def send_message(self, mess):
         super().send_message(mess)
-        logging.debug("trying to send to node %s" % mess.to.node)
-        logging.debug("trying to send to resource %s" % mess.to.resource)
-        logging.debug("trying to type %s" % mess.type)
-        if mess.type == 'groupchat':
-            logging.debug('send grouchat message to %s' % mess.to.resource)
-            self.sc.rtm_send_message(mess.to.node, mess.body)
-        else:
-            logging.debug('send chat message to %s' % mess.to.resource)
-            self.sc.rtm_send_message(mess.to.node, mess.body)
-
+        to_humanreadable = "<unknown>"
+        try:
+            if mess.type == 'groupchat':
+                to_humanreadable = mess.to.domain
+                to_id = self.channelname_to_channelid(to_humanreadable)
+            else:
+                to_humanreadable = mess.to.node
+                api_data = api_resp(
+                    self.sc.api_call(
+                        'im.open',
+                        user=self.username_to_userid(to_humanreadable)
+                    )
+                )
```

Comment:
There's a TODO here. Since we never close the IM sessions, we should cache the result of this API call so we only have to do it once for each user, thereby saving a whole bunch of API calls (that count towards our rate limits).